### PR TITLE
improve javadoc for LLC and mgmt clients

### PR DESCRIPTION
**Pull request by XiaofeiCao**
fix https://github.com/Azure/autorest.java/issues/1178

This PR mainly fixes javadoc on `Mono` and `Response` return types:

## Code change
Changes are mainly in [ReturnTypeDescriptionAssembler](https://github.com/Azure/autorest.java/pull/1299/files#diff-e58e4a1463f332292bddc0f52d3ac96ac045cfd3def72bac55bd8eed79c2bed3) and [ClientMethodMapper](https://github.com/Azure/autorest.java/pull/1299/files#diff-e8565e9338fcb9779250f4a448d1964afd7c5d6336316082f0f33579e92e3eb5)

### For `Mono`:
* `Mono<Void>` - A {@link Mono} that completes when a successful response is received
https://github.com/Azure/autorest.java/blob/b26b2df852e29e344a4367bc73068ef2b74f0044/azure-tests/src/main/java/fixtures/azureparametergrouping/ParameterGroupings.java#L221-L229
* `Mono<Response>` - Return `Response description` on successful completion of {@link Mono} (`Response description` here, is the description of the generic type argument, which is Response)
https://github.com/Azure/autorest.java/blob/b26b2df852e29e344a4367bc73068ef2b74f0044/azure-tests/src/main/java/fixtures/azureparametergrouping/ParameterGroupings.java#L260-L271
* `Mono<T>` - Return "something" on successful completion of {@link Mono} (something here is the description in the operation or in the response schema)
https://github.com/Azure/autorest.java/blob/b26b2df852e29e344a4367bc73068ef2b74f0044/customization-tests/src/main/java/fixtures/bodycomplex/implementation/InheritancesImpl.java#L96-L103
* `Mono<OtherType>` - Returns the response body on successful completion of {@link Mono} (OtherType is the situation where no description is in the operation or response schema)
https://github.com/Azure/autorest.java/blob/b26b2df852e29e344a4367bc73068ef2b74f0044/customization-tests/src/main/java/fixtures/bodycomplex/implementation/FlattencomplexesImpl.java#L82-L88

## for `Response`:
* `Response<Void>` - Returns the {@link Response}
https://github.com/Azure/autorest.java/blob/4c600510ab71475fc1588dcf483d566159dc101d/azure-tests/src/main/java/fixtures/azurespecials/Odatas.java#L180-L193
* `Response<T>` - Returns "something" along with {@link Response}
https://github.com/Azure/autorest.java/blob/4c600510ab71475fc1588dcf483d566159dc101d/customization-tests/src/main/java/fixtures/bodycomplex/implementation/ArraysImpl.java#L170-L180
* `Response<OtherType>` - Returns the response body along with {@link Response} (OtherType is the situation where no description is in the operation or response schema)
https://github.com/Azure/autorest.java/blob/4c600510ab71475fc1588dcf483d566159dc101d/customization-tests/src/main/java/fixtures/bodycomplex/PolymorphismClient.java#L270-L282

### CODE REVIEW

File: javagen/src/main/java/com/azure/autorest/mapper/ClientMethodMapper.java [hunk 1]
```diff
@@ -894,21 +896,19 @@ protected static String returnTypeDescription(Operation operation, IType returnT
...
-                }
+            if (description == null
+                && baseType == PrimitiveType.Boolean
+                && operation.getRequests() != null && !operation.getRequests().isEmpty()
+                && operation.getRequests().get(0).getProtocol() != null
+                && operation.getRequests().get(0).getProtocol().getHttp() != null
+                && HttpMethod.HEAD.name().equalsIgnoreCase(operation.getRequests().get(0).getProtocol().getHttp().getMethod())) {
+                // Mono<Boolean> of HEAD method
+                description = "whether resource exists";
+                return description;
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/mapper/ClientMethodMapper.java:907]: Personally I would try avoid `return` in middle of method. It is kind of `goto`, and make the logic hard to understand.

File: javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
+package com.azure.autorest.util.returntype;
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java:1]: copyright

File: javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        @Override
+        public String handle(String description, GenericType returnType, IType baseType) {
+            if (description == null) {
+                if (ClassType.Void == baseType) {
+                    return "A {@link Mono} that completes when a successful response is received.";
+                } else {
+                    return "the response body on successful completion of {@link Mono}";
+                }
+            } else {
+                return String.format("%s on successful completion of {@link Mono}", description);
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java:56]: To be consistent on the period at the end of the sentence.

File: javagen/src/test/java/com/azure/autorest/JavagenUnitTests.java [hunk 4]
```diff
@@ -31,4 +31,23 @@ public void generateTest(){
         System.out.println(javaPackage.getJavaFiles().size());
     }
 
+    @Test
+    public void generateWebpubsubTest(){
+        MockUnitJavagen javagen = new MockUnitJavagen();
+        // SOURCE SWAGGER URL: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/webpubsub/data-plane/WebPubSub/stable/2021-10-01/webpubsub.json
+        // It's generated by preprocessor. It can be found under preprocessor module after calling autorest for generation.
+        String fileName = "webpubsub-code-model-processed-no-tags.yaml";
```
Comment (weidongxu-microsoft) [inline javagen/src/test/java/com/azure/autorest/JavagenUnitTests.java:39]: Reason about needing a new test input?

Comment (XiaofeiCao) [reply javagen/src/test/java/com/azure/autorest/JavagenUnitTests.java:39]: I wanted to test the javadoc from various swagger inputs. It's kind of like an experiment.

**Comment by weidongxu-microsoft** [comment]
Run `generate.bat` or `generate` after you've done all the coding, to refresh to code for autorest.testserver.

File: azure-tests/src/main/java/fixtures/azureparametergrouping/ParameterGroupings.java [hunk 5]
```diff
@@ -223,7 +223,7 @@ public void postRequired(ParameterGroupingPostRequiredParameters parameterGroupi
      *
      * @throws ErrorException thrown if the request is rejected by server.
      * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
-     * @return the completion.
+     * @return A {@link Mono} that completes when a successful response is received.
```
Comment (XiaofeiCao) [inline azure-tests/src/main/java/fixtures/azureparametergrouping/ParameterGroupings.java:226]: Mono<Void>

File: azure-tests/src/main/java/fixtures/azureparametergrouping/ParameterGroupings.java [hunk 6]
```diff
@@ -264,7 +264,7 @@ public void postOptional() {
      * @throws IllegalArgumentException thrown if parameters fail the validation.
      * @throws ErrorException thrown if the request is rejected by server.
      * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
-     * @return the completion.
+     * @return the {@link Response} on successful completion of {@link Mono}.
```
Comment (XiaofeiCao) [inline azure-tests/src/main/java/fixtures/azureparametergrouping/ParameterGroupings.java:267]: `Mono<Response>`

File: customization-tests/src/main/java/fixtures/bodycomplex/implementation/InheritancesImpl.java [hunk 7]
```diff
@@ -98,7 +98,7 @@
      *
      * @throws ErrorException thrown if the request is rejected by server.
      * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
-     * @return complex types that extend others.
+     * @return complex types that extend others on successful completion of {@link Mono}.
```
Comment (XiaofeiCao) [inline customization-tests/src/main/java/fixtures/bodycomplex/implementation/InheritancesImpl.java:101]: `Mono<T>`

File: customization-tests/src/main/java/fixtures/bodycomplex/implementation/FlattencomplexesImpl.java [hunk 8]
```diff
@@ -82,7 +82,7 @@
     /**
      * @throws HttpResponseException thrown if the request is rejected by server.
      * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
-     * @return the response.
+     * @return the response body on successful completion of {@link Mono}.
```
Comment (XiaofeiCao) [inline customization-tests/src/main/java/fixtures/bodycomplex/implementation/FlattencomplexesImpl.java:85]: `Mono<OtherType>`

File: javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+import com.azure.autorest.model.clientmodel.PrimitiveType;
+import com.azure.core.http.rest.Response;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Mono;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ReturnTypeDescriptionAssembler {
+    private static final ClientLogger LOGGER = new ClientLogger(ReturnTypeDescriptionAssembler.class);
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java:20]: Use the `PluginLogger` instead.

autorest.java integrates with autorest, and had to use its logging system.

Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java]: Switched to `PluginLogger`. Thanks for pointing out!

File: azure-tests/src/main/java/fixtures/azurespecials/Odatas.java [hunk 10]
```diff
@@ -187,7 +187,7 @@ public void getWithFilter() {
      * @throws IllegalArgumentException thrown if parameters fail the validation.
      * @throws ErrorException thrown if the request is rejected by server.
      * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
-     * @return the response.
+     * @return the {@link Response}.
```
Comment (XiaofeiCao) [inline azure-tests/src/main/java/fixtures/azurespecials/Odatas.java:190]: `Response<Void>`

File: customization-tests/src/main/java/fixtures/bodycomplex/implementation/ArraysImpl.java [hunk 11]
```diff
@@ -174,7 +174,7 @@ public ArrayWrapper getValid() {
      * @throws IllegalArgumentException thrown if parameters fail the validation.
      * @throws ErrorException thrown if the request is rejected by server.
      * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
-     * @return complex types with array property.
+     * @return complex types with array property along with {@link Response}.
```
Comment (XiaofeiCao) [inline customization-tests/src/main/java/fixtures/bodycomplex/implementation/ArraysImpl.java:177]: `Response<T>`

File: customization-tests/src/main/java/fixtures/bodycomplex/PolymorphismClient.java [hunk 12]
```diff
@@ -275,7 +275,7 @@ public Salmon putMissingDiscriminator(Salmon complexBody) {
      * @throws IllegalArgumentException thrown if parameters fail the validation.
      * @throws ErrorException thrown if the request is rejected by server.
      * @throws RuntimeException all other wrapped checked exceptions if the request fails to be sent.
-     * @return the response.
+     * @return the response body along with {@link Response}.
```
Comment (XiaofeiCao) [inline customization-tests/src/main/java/fixtures/bodycomplex/PolymorphismClient.java:278]: `Response<OtherType>`

File: javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java [hunk 13]
```diff
@@ -100,7 +100,7 @@ public boolean accept(IType returnType) {
         public String handle(String description, GenericType returnType, IType baseType) {
             String assembledDesc;
             if (description == null) {
-                if (PrimitiveType.Void == baseType) { // Response<Void>
+                if (PrimitiveType.Void == baseType || PrimitiveType.Void == returnType.getTypeArguments()[0]) { // Response<Void>
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java:103]: The reason adding `PrimitiveType.Void == returnType.getTypeArguments()[0]` condition? And why it is not `ClassType.Void`?

File: javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java [hunk 14]
```diff
@@ -0,0 +1,143 @@
...
+        }
+    }
+
+    private Class<?> getGenericClass(GenericType type) {
+        String className = String.format("%s.%s", type.getPackage(), type.getName());
+        try {
+            return Class.forName(className);
+        } catch (ClassNotFoundException e) {
+            logger.error(String.format("class %s not found!", className), e);
+            throw new RuntimeException(e);
```
Comment (weidongxu-microsoft) [inline javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java:121]: It occurred to me that this is codegen, it does not really had to load all the classes that to be generated in runtime.

So you should not throw exception here. If class not found in runtime, you know it is unknown, log a warning, and that's all.

File: javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java (hunk 13 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java]: In situations like `create simple async rest response`, `syncReturnType` is `PrimitiveType.Void` and is passed in as `baseType` parameter. `ClassType.Void` is more semantically suitable for `baseType` here and I added `baseType.asNullable()` to handle this situation.

**Comment by XiaofeiCao** [comment]
@srnagar 
Hi, please take a look at this PR of improvement on returnType javadoc, and comment for any suggestions, thanks!

**Review by weidongxu-microsoft — APPROVED** [review]

**Review by weidongxu-microsoft — APPROVED** [review]

File: javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java [hunk 15]
```diff
@@ -0,0 +1,148 @@
...
+import com.azure.autorest.model.clientmodel.ClassType;
+import com.azure.autorest.model.clientmodel.GenericType;
+import com.azure.autorest.model.clientmodel.IType;
+import com.azure.core.http.rest.Response;
+import reactor.core.publisher.Mono;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class ReturnTypeDescriptionAssembler {
```
Comment (srnagar) [inline javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java:20]: nit: Add javadoc to describe the purpose of this class.

File: javagen/src/test/resources/webpubsub-code-model-processed-no-tags.yaml [hunk 16]
```diff
@@ -0,0 +1,3381 @@
+&id003 !!com.azure.autorest.extension.base.model.codemodel.CodeModel
```
Comment (srnagar) [inline javagen/src/test/resources/webpubsub-code-model-processed-no-tags.yaml]: This file should not be checked in.

File: javagen/src/test/java/com/azure/autorest/mapper/ClientMethodMapperAccessor.java [hunk 17]
```diff
@@ -0,0 +1,16 @@
+package com.azure.autorest.mapper;
+
+import com.azure.autorest.extension.base.model.codemodel.Operation;
+import com.azure.autorest.model.clientmodel.IType;
+
+/**
+ * @author xiaofeicao
+ * @createdAt 2021-12-29 15:14
+ */
+public class ClientMethodMapperAccessor extends ClientMethodMapper{
```
Comment (srnagar) [inline javagen/src/test/java/com/azure/autorest/mapper/ClientMethodMapperAccessor.java:10]: Why do we need to extend the `ClientMethodMapper` to have a static method in it?

File: javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java [hunk 18]
```diff
@@ -0,0 +1,148 @@
...
+    private final PluginLogger logger;
+    private final ReturnTypeDescriptionHandlerRegistry handlerRegistry;
+
+    public ReturnTypeDescriptionAssembler(NewPlugin host) {
+        logger = new PluginLogger(host, ReturnTypeDescriptionAssembler.class);
+        handlerRegistry = new ReturnTypeDescriptionHandlerRegistry();
+        handlerRegistry.addFirst(new MonoDescriptionHandler());
```
Comment (srnagar) [inline javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java:27]: Do we need the complexity of a registry to handle a couple of return types? Can this be simplified in the ClientMethodTemplate to add the description based on return type?

File: javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java [hunk 19]
```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ */
+
+
+package com.azure.autorest.util.returntype;
```
Comment (srnagar) [inline javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java:7]: Do we need a separate package specific to `returntype`?

File: javagen/src/test/java/com/azure/autorest/mapper/ClientMethodMapperAccessor.java (hunk 17 above)
Comment (XiaofeiCao) [reply javagen/src/test/java/com/azure/autorest/mapper/ClientMethodMapperAccessor.java:10]: It's due to test case invocation here https://github.com/XiaofeiCao/autorest.java/blob/2136e157a43171f7375db7e83ec3aff96a90324d/javagen/src/test/java/com/azure/autorest/JavagenUnitTests.java#L78
Original method is `protected` so I subclassed to create a `public` method which invokes the original method, to allow for test case invocation.

File: javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java (hunk 19 above)
Comment (XiaofeiCao) [reply javagen/src/main/java/com/azure/autorest/util/returntype/ReturnTypeDescriptionAssembler.java:7]: I'll move it to an existing package in the next PR.

**Pull request merged**